From 2c5e13b8eaacf36cf2223fc7ce79fd9711307b91 Mon Sep 17 00:00:00 2001
From: Hai Pham <hai.pham.ud@renesas.com>
Date: Wed, 7 May 2025 11:47:55 +0700
Subject: [PATCH 31/52] i2c: rcar_i2c: Use fixed clock value for R-Car Gen5

Use fixed clock value for R-Car Gen5 I2C, with SCPBUSD4_SCP_MAIN at 150Mhz for
channel 0 and SGASYNCD8_PERW_MAIN at 133.33Mhz for remaining channels (1-8)

Signed-off-by: Hai Pham <hai.pham.ud@renesas.com>
---
 drivers/i2c/rcar_i2c.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/drivers/i2c/rcar_i2c.c b/drivers/i2c/rcar_i2c.c
index 9562119bb5b..dcbe0b0c208 100644
--- a/drivers/i2c/rcar_i2c.c
+++ b/drivers/i2c/rcar_i2c.c
@@ -54,6 +54,8 @@
 #define RCAR_I2C_ICFBSCR		0x38
 #define RCAR_I2C_ICFBSCR_TCYC17		0x0f /* 17*Tcyc */
 
+#define CFG_CLK_IGNORE
+
 enum rcar_i2c_type {
 	RCAR_I2C_TYPE_GEN2,
 	RCAR_I2C_TYPE_GEN3,
@@ -262,7 +264,11 @@ static int rcar_i2c_set_speed(struct udevice *dev, uint bus_freq_hz)
 	 * clkp : peripheral_clk
 	 * F[]  : integer up-valuation
 	 */
+#if !defined(CFG_CLK_IGNORE)
 	rate = clk_get_rate(&priv->clk);
+#else
+	rate = (uintptr_t)priv->base == 0xc11d0000 ? 150000000 : 133333333;
+#endif
 	cdf = rate / 20000000;
 	if (cdf >= 8) {
 		dev_err(dev, "Input clock %lu too high\n", rate);
@@ -304,7 +310,7 @@ static int rcar_i2c_set_speed(struct udevice *dev, uint bus_freq_hz)
 
 scgd_find:
 	dev_dbg(dev, "clk %d/%d(%lu), round %u, CDF:0x%x, SCGD: 0x%x\n",
-		scl, bus_freq_hz, clk_get_rate(&priv->clk), round, cdf, scgd);
+		scl, bus_freq_hz, rate, round, cdf, scgd);
 
 	priv->icccr = (scgd << RCAR_I2C_ICCCR_SCGD_OFF) | cdf;
 	writel(priv->icccr, priv->base + RCAR_I2C_ICCCR);
@@ -331,6 +337,7 @@ static int rcar_i2c_probe(struct udevice *dev)
 					      "i2c-scl-internal-delay-ns", 5);
 	priv->type = dev_get_driver_data(dev);
 
+#if !defined(CFG_CLK_IGNORE)
 	ret = clk_get_by_index(dev, 0, &priv->clk);
 	if (ret)
 		return ret;
@@ -338,6 +345,7 @@ static int rcar_i2c_probe(struct udevice *dev)
 	ret = clk_enable(&priv->clk);
 	if (ret)
 		return ret;
+#endif
 
 	/* reset slave mode */
 	writel(0, priv->base + RCAR_I2C_ICSIER);
@@ -352,8 +360,10 @@ static int rcar_i2c_probe(struct udevice *dev)
 	writel(0, priv->base + RCAR_I2C_ICMAR);
 
 	ret = rcar_i2c_set_speed(dev, I2C_SPEED_STANDARD_RATE);
+#if !defined(CFG_CLK_IGNORE)
 	if (ret)
 		clk_disable(&priv->clk);
+#endif
 
 	return ret;
 }
-- 
2.34.1

